src/cargo/util/semver_ext.rs

```diff
@@ -84,10 +84,17 @@ impl OptVersionReq {
             OptVersionReq::Any => true,
             OptVersionReq::Req(req) => req.matches(version),
             OptVersionReq::Locked(v, _) => {
-                v.major == version.major
-                    && v.minor == version.minor
-                    && v.patch == version.patch
-                    && v.pre == version.pre
+                // We wuld like to use `version == v`.
+                // Generally  we can, lockfiles record an exact version
+                // which is the only one we should use on repeat builds.
+                // However, this is also used for the `--precise` field of cargo update
+                // in that context we treat a requirement that does not have build metadata as allowing any metadata.
+                version.major == v.major
+                    && version.minor == v.minor
+                    && version.patch == v.patch
+                    && version.pre == v.pre
+                    && (version.build == v.build
+                        || (!version.build.is_empty() && v.build.is_empty()))
```

Comment:
Can't this just be `|| v.build.is_empty()` since if they are both empty, that will be caught in the `==`.

I think I'd also find that easier to read.  For me, seeing just "v.build.is_empty()`, that makes it sounds like "this is a wild card" but seeing it `&&` with something else makes it seem like something more is going on